### PR TITLE
View specialization

### PR DESCRIPTION
**Pull request by shazqadeer**
This PR adds specialization shortcuts for converting n-d to 1-d and 1-d to 2-d views. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: test/export/test_export.py
Comment (shazqadeer) [inline test/export/test_export.py:3312]: @ezyang : I have run into an interesting conundrum with this test case ```test_redundant_asserts```. This test, along with all the other tests in this file, are reused with some changes (that I don't quite understand) from many other files in this folder such as test_export_nonstrict.py, test_retraceability.py, test_serdes.py, etc. It turns out that the constraint ```torch._check(a.numel() > 0)``` in my specialization functions gets added sometimes but not always in the generated code over which the count of "torch.ops.aten._assert_scalar.default" is being performed. So an exact match of either 1 or 2 count does not work across the situations in which this test is being reused. What do you recommend?

Comment (ezyang) [reply test/export/test_export.py:3312]: @tugsbayasgalan would you be able to advise?

Sometimes, there is a combinator that lets you vary the expected count on a config per config basis, but I don't see any occurrence of this in this particular test file. Another possibility is that the test is written poorly and you need to write the test in a more generic way that works for the situation.

However, I observe one more very important thing. You are testing if `a.numel() > 0`. But what if a has no elements?!

File: torch/_refs/__init__.py
Comment (ezyang) [inline torch/_refs/__init__.py:3674]: There's no point in guard_size_oblivious on ndim, dimensionality of tensors cannot be dynamic

File: torch/_refs/__init__.py
Comment (ezyang) [inline torch/_refs/__init__.py:3619]: There's no reason to exclude 0-d scalar tensors here.

```
>>> torch.randn([]).view(1)
tensor([-1.5574])
```

File: torch/_refs/__init__.py
Comment (ezyang) [inline torch/_refs/__init__.py:3629]: This is not quite right. You need to account for when dim0 or dim1 == -1

File: torch/_refs/__init__.py
Comment (ezyang) [inline torch/_refs/__init__.py:3624]: Since this file has a lot of other functions in it, you need to be more specific about what exactly these functions do, or move them into view_reshape_helper, or just don't bother factoring them as helpers at all

File: torch/_refs/__init__.py
Comment (ezyang) [inline torch/_refs/__init__.py:3621]: Need to handle non-zero storage offset here (and below) correctly

File: torch/_refs/__init__.py
Comment (ezyang) [reply torch/_refs/__init__.py:3627]: It is good to distinguish them (or ideally, NOT test the preconditions at all), because some preconditions are user error, and other are "oops the caller code shouldn't have called this helper", and they're all jumbled together right now

File: test/export/test_export.py
Comment (shazqadeer) [reply test/export/test_export.py:3312]: The condition ```a.numel() == 0``` is handled before the call to these functions in ```_reshape_view_helper```.  See line 3641.

File: torch/_refs/__init__.py
Comment (shazqadeer) [reply torch/_refs/__init__.py:3674]: Good point. Will fix.

File: torch/_refs/__init__.py
Comment (shazqadeer) [reply torch/_refs/__init__.py:3629]: I believe the inference is being done in ```_reshape_view_helper``` before the call to these functions. See line 3638.

I am going to add comments to these functions.

File: torch/_refs/__init__.py
Comment (shazqadeer) [reply torch/_refs/__init__.py:3621]: Hmm. I read this [documentation](https://pytorch.org/docs/stable/generated/torch.as_strided.html#torch-as-strided) which suggests that if I don't explicitly specify the optional storage_offset, then the storage_offset for the output will match that of the input. Is that not what we want?

File: torch/_refs/__init__.py
Comment (shazqadeer) [reply torch/_refs/__init__.py:3624]: I agree.

File: test/export/test_export.py
Comment (ezyang) [reply test/export/test_export.py:3312]: Oh, well, that explains why sometimes you end up with two. `a.numel() > 0` is different from `a.numel() == 0`, and additionally, the latter test was done under `guard_size_oblivious`. So it might be the case that `a.numel() == 0` when size oblivious, but NOT without it, which is why `a.numel() > 0` results in an extra guard. Really, you should just delete this assert entirely...

File: torch/_refs/__init__.py
Comment (ezyang) [reply torch/_refs/__init__.py:3629]: oh oops, you're right. But then do you even need these asserts?

File: torch/_refs/__init__.py
Comment (shazqadeer) [reply torch/_refs/__init__.py]: I have eliminated the separate functions which seem unnecessary now. Hopefully, the revised code resolves some of these concerns.

File: torch/_refs/__init__.py
Comment (shazqadeer) [reply torch/_refs/__init__.py]: I have removed this guard. Note that the guard is true though because of the code in ```reshape_view_helper``` before getting here.

File: torch/_refs/__init__.py
Comment (shazqadeer) [reply torch/_refs/__init__.py]: I wrote them more as "preconditions" i.e., facts which must be guaranteed by preceding code. I don't think I need them for solving Ivan's problem. I will remove them.

File: test/export/test_export.py
Comment (shazqadeer) [reply test/export/test_export.py:3312]: I tried deleted the ```a.numel() > 0``` but then Ivan's test does not work.

Comment (ezyang) [reply test/export/test_export.py]: That seems... odd? Do you have a minimized version of the test that you can put directly into the test suite here?

Comment (shazqadeer) [reply test/export/test_export.py]: Yes, I am going to take a deeper look at that failure. I was also puzzled because of the earlier test of ```guard_size_oblivious(a.numel() == 0)```. With that in place, I assumed that if the function had not returned earlier, ```a.numel() > 0``` would be in the context. 

I am wondering if the strengthening that happens inside ```guard_size_oblivious``` corresponding causes a weakening when the Boolean condition is negated to follow control down the other path.

**Comment by shazqadeer** [comment]
@JackCaoG : An XLA test is failing because presumably GT (>) is not implemented. Any suggestions to unblock this PR?

**Comment by JackCaoG** [comment]
The test is in XLA I think, we need to implement https://github.com/pytorch/xla/blob/94fa999380d51ca79ee3b3c8546fd49ce6815298/torch_xla/csrc/tensor.cpp#L741-L744.

@vanbasten23 do you think this will be a quick lowering?

**Comment by shazqadeer** [comment]
@JackCaoG : Thanks for the quick response. In the meantime, I am trying to see if ```!=``` works instead of ```>```. In my code, the two comparisons have equivalent semantics.

File: torch/_refs/__init__.py
Comment (shazqadeer) [inline torch/_refs/__init__.py:3652]: @JackCaoG , @ezyang : I tried to avoid taking the specialized path by adding the condition ```a.device.type != "xla"```. But it appears the xla test is still failing on line 3653. Is the condition incorrect?

Comment (JackCaoG) [reply torch/_refs/__init__.py:3652]: yea it seems right to me
```
>>> import torch
>>> import torch_xla
>>> t1 = torch.randn(3,3, device=torch_xla.device())
>>> t1.device.type
'xla'
>>> t1.device.type == "xla"
True
```

Comment (shazqadeer) [reply torch/_refs/__init__.py:3652]: Hmm, then how to I make sense of this XLA test failure:

https://hud.pytorch.org/pr/pytorch/pytorch/127641#25765362668

File: torch/_refs/__init__.py
Comment (IvanKobzarev) [inline torch/_refs/__init__.py:3653]: Current eager allows to work with 0 numel tensors in the way, we should allow a.numel() == 0
e.g.
```
import torch
t = torch.tensor([])
t.shape
>> torch.Size([0])
t.numel()
>> 0
r = t.view(0, 10)
r.shape
>> torch.Size([0, 10])
```

Comment (shazqadeer) [reply torch/_refs/__init__.py:3653]: The case ```a.numel() == 0``` is already covered earlier in the code on line 3627.

**Comment by shazqadeer** [comment]
I have eliminated the check ```a.numel() > 0``` which has been the source of so much friction.

**Comment by shazqadeer** [comment]
I see this failure which does not have anything to do with my build:
```
manifest for nvidia/cuda:11.8-cudnn9-devel-ubuntu20.04 not found: manifest unknown: manifest unknown
```

Anybody seen this before?

**Comment by shazqadeer** [comment]
@ezyang : I think this PR is ready to go in. I don't know about these build failures. They failed in the morning also. So I am not sure any more if they are just flaky failures.

File: test/export/test_export.py
Comment (ezyang) [inline test/export/test_export.py:166]: nice

File: test/dynamo/test_view.py
Comment (ezyang) [inline test/dynamo/test_view.py:7]: Don't do this, this will interfere with other tests if they're running same process (as is the case if you do pytest.) Instead, use the decorator `@torch._dynamo.config.patch` to temporarily change config. Or if you put these tests in test_unspec.py, the class globally has a patch for all tests in that file.

**Review by ezyang — APPROVED** [review]
Please don't land without fixing the test pollution problem

File: test/dynamo/test_view.py
Comment (shazqadeer) [reply test/dynamo/test_view.py]: Fixed. Please take a look.

**Comment by ezyang** [comment]
Failure looks real

**Pull request closed without merging**
